Create redis command router module

StFS opened the pull request:
This is a DRAFT PR to add a Redis backed command-router.

It consists of the following changes:

- Adds a `client-device-connection-redis` module that implements a `device-connection-info` repository using the Quarkus Redis Client extension.
- Adds a `services/command-router-redis` module that creates a `command-router-redis` Docker image.
- Adds the necessary configuration to the integration test project to start up a Redis server beneath the new Redis backed Command Router service (profile `redis_cache`).

This PR was delayed by sporadic failures of `CommandAndControlAmqpIT`/`CommandAndControlMqttIT`. After investigation, they turned out to be two independent client-side defects:

1. Connection-pool rejection under command bursts. The vert.x Redis client acquires a pooled connection per command, so one Kafka poll batch of inbound commands (each needing 1–2 `MGET` lookups) could exceed the default pool capacity (6 connections, wait queue 24). Excess acquisitions failed with `ConnectionPoolTooBusyException`, silently dropping command delivery (confirmed via `redis_pool_rejected_total` and Jaeger traces). Fixed by shipping `max-pool-waiting: 512` as a module default - a deep wait queue smooths the burst at a few milliseconds of queueing latency instead of dropping commands.
2. Pooled-connection poisoning by `WATCH`/`MULTI`. The vert.x Redis client returns pooled connections without resetting transaction state (its taint tracking only covers `pub/sub`, `SELECT` and `AUTH`). `remove()` leaked an armed `WATCH` on its common early-return paths and `putAll(…, lifespan)` could leak an open `MULTI`, silently corrupting unrelated later operations that borrowed the same connection - including transaction aborts that were reported as success. This matches the observed "after one failure, subsequent runs keep failing" pattern. Fixed by removing connection state entirely: `remove()` and `putAll-with-TTL` are now atomic server-side Lua scripts executed via `EVAL` over the pooled RedisAPI. An upstream issue against vertx-redis-client may get filed separately.

Related correctness fixes that surfaced when addressing that: `get()` now succeeds with null for missing keys per the Cache contract (previously an NPE-failed future turned last-known-gateway 404s into 500s), transaction aborts can no longer be silently swallowed (no transactions remain), and empty-map `putAll` is guarded.

Requirements note: the module now requires Lua scripting (`EVAL`, Redis ≥ 2.6). When Redis ACLs deny `@scripting`, startup fails fast with a descriptive, logged error (see the module README).

Testing

- New `RedisCacheTest` (13 tests, Testcontainers, real Redis 8): full Cache contract coverage, run with `max-pool-size = 1` so any reintroduced connection-state leak fails the suite, including a dedicated regression test that reproduced the poisoning sequence before the fix.
- Validated end-to-end: 12 consecutive loop iterations of both `CommandAndControl` suites against the built images - 12/12 green, `redis_pool_rejected_total = 0` and zero failed Redis commands across the entire window (historical baseline failed within 1–6 iterations).
- Redis pool metrics (`redis_pool_*`) are now exposed (`quarkus.micrometer.binder.redis.enabled` plus enumerating the relocated `hono.commandRouter.cache.redis.*` names in `RedisConfigInterceptor`), which is what made both root causes observable.

Observation for reviewers: `CommandRouterServiceImpl.start()` currently discards the cache lifecycle future, so startup failures of any cache backend are invisible to the service lifecycle; the Redis cache therefore also logs its startup failure itself. Possibly worth addressing in `command-router-base` separately.

StFS committed:
Add Redis backed Command Router by adding a client-device-connection-redis module and a services/command-router-redis module. Also add appropriate configurations to integration tests to run them against a Redis backed command router with a Redis server

StFS committed:
Add Redis backed Command Router by adding a client-device-connection-redis module and a services/command-router-redis module. Also add appropriate configurations to integration tests to run them against a Redis backed command router with a Redis server

StFS committed:
Update dependencies and clean up pom.xml files after syncing to latest Hono version

StFS committed:
Update dependencies and clean up pom.xml files after syncing to latest Hono version

StFS commented:
Addresses https://github.com/eclipse-hono/hono/issues/3532

StFS commented:
This comment is a reminder to myself.

To run the integration tests against the redis command router:

```
mvn clean verify -Prun-tests,redis_cache
```

Currently I'm trying to run only the tests that fail... but they only fail when I run the whole integration test suite. If I limit the run to the tests that fail the tests don't fail. Here's the command I'm using to run only the failing tests:

```
mvn clean verify -Prun-tests,redis_cache -Dit.test=CommandAndControlAmqpIT -Dit.test=CommandAndControlMqttIT
```

So what I'm going to try now is to run the tests once using the `-Ddocker.keepRunning` option and then repeatedly run the tests that fail against the containers that were started until a run fails (if that happens). Here are the commands I'm using for this:

```
# Run once to start all the containers
mvn clean verify -Prun-tests,redis_cache -Dit.test=CommandAndControlAmqpIT -Dit.test=CommandAndControlMqttIT -Ddocker.keepRunning

# Run the tests repeatedly until a run fails
while ../mvnw verify -Prun-tests,redis_cache,useRunningContainers -Dit.test=CommandAndControlAmqpIT -Dit.test=CommandAndControlMqttIT ; do :; done
```

StFS commented:
I have verified that when I run the sequence described above, in the while loop that repeatedly runs the tests, eventually the tests will fail with a timeout error (on the test run side). In the command router service logs there are log lines that display some warnings related to the Kafka client there. These log lines and warnings also seem to appear in the infinispan command router when I run the tests against that, but that does not result in the same timeout error. However, in the redis command router, when the tests error out, I did find the following error in the command router logs which does not appear in the logs for the infinispan based command router:

```
2025-01-18 03:10:32,855 WARN  [org.apa.kaf.cli.NetworkClient] (kafka-producer-network-thread | event-bde4a74d-3f49-4e1c-b1e8-f1585409253e) [Producer clientId=event-bde4a74d-3f49-4e1c-b1e8-f1585409253e] Error while fetching metadata with correlation id 193 : {hono.event.a530c0a3-2ca0-4055-b360-02c8b892f387=UNKNOWN_TOPIC_OR_PARTITION, hono.event.1cd1fb42-bb19-45e8-8e55-41e335b36deb=UNKNOWN_TOPIC_OR_PARTITION, hono.event.a44a8fb0-3e9b-4359-a4a4-502ec57bf3b4=UNKNOWN_TOPIC_OR_PARTITION, hono.event.ef07c6f2-2fb7-4ff3-93af-9f8b27906b49=UNKNOWN_TOPIC_OR_PARTITION, hono.event.942dea39-14e4-47bf-9cdf-d4cd83cd48b5=UNKNOWN_TOPIC_OR_PARTITION, hono.event.febe7482-629e-487f-8b79-3625db931d9e=UNKNOWN_TOPIC_OR_PARTITION, hono.event.9721c210-c90a-40a0-80c9-8b5f5ef10589=UNKNOWN_TOPIC_OR_PARTITION, hono.event.b81bf5d2-b9e5-40a9-9df4-3cbc5e24c5b5=UNKNOWN_TOPIC_OR_PARTITION, hono.event.14ac4593-fa4f-4d5d-a1d5-0d7c272d6033=UNKNOWN_TOPIC_OR_PARTITION, hono.event.d12ff377-565a-404a-a1b1-40aece6a089e=UNKNOWN_TOPIC_OR_PARTITION, hono.event.44cf0462-e3f4-4256-8124-b3cde64999ac=UNKNOWN_TOPIC_OR_PARTITION, hono.event.5b4e4805-e8f7-4180-9b8b-fb9b513a43f0=UNKNOWN_TOPIC_OR_PARTITION, hono.event.2b51dfc8-6b00-4f64-b241-aa15a0e58e5e=UNKNOWN_TOPIC_OR_PARTITION, hono.event.0a14698e-a3be-4ac2-99e9-4616e946de0d=UNKNOWN_TOPIC_OR_PARTITION, hono.event.8de73437-63f4-4723-b647-82a3ad0d58b9=UNKNOWN_TOPIC_OR_PARTITION, hono.event.21ad0229-a364-4d9e-9e85-3883e65034a3=UNKNOWN_TOPIC_OR_PARTITION, hono.event.8803b596-dc36-4a64-b4f0-ba01aa9903c2=UNKNOWN_TOPIC_OR_PARTITION, hono.event.9ee42165-e83e-4369-91e2-7978a6b07800=UNKNOWN_TOPIC_OR_PARTITION, hono.event.99ab674f-63e0-4521-890e-2d6fced95d36=UNKNOWN_TOPIC_OR_PARTITION, hono.event.0661fc86-2b08-41f6-9fbf-dd831c6ad5fd=UNKNOWN_TOPIC_OR_PARTITION, hono.event.60160670-f222-4e5f-99fe-7928063d0bba=UNKNOWN_TOPIC_OR_PARTITION, hono.event.2633c7ec-152d-448e-9dd0-a5617b2f40ed=UNKNOWN_TOPIC_OR_PARTITION, hono.event.b9f90aab-dc4c-4b7b-b03e-fb0c0a624839=UNKNOWN_TOPIC_OR_PARTITION, hono.event.0660782c-7317-43b1-a71c-af1c0b39131f=UNKNOWN_TOPIC_OR_PARTITION, hono.event.a838f1dc-0009-4354-aff5-261ebd23eb01=UNKNOWN_TOPIC_OR_PARTITION, hono.event.b1e56267-da24-4fa4-a996-6a5a3aa4ac9b=UNKNOWN_TOPIC_OR_PARTITION, hono.event.d37ca8b0-af8d-4393-8c32-61be503da711=UNKNOWN_TOPIC_OR_PARTITION, hono.event.d8ee36b0-b49c-4849-8510-f922a414c2ea=UNKNOWN_TOPIC_OR_PARTITION, hono.event.4889160b-cf0e-4bed-b987-9f07983affc4=UNKNOWN_TOPIC_OR_PARTITION, hono.event.ff972a2d-f300-41ba-bb1a-c48fbde40163=UNKNOWN_TOPIC_OR_PARTITION, hono.event.1a8afec8-8354-40ea-aabe-367e13452fc0=UNKNOWN_TOPIC_OR_PARTITION, hono.event.51800a66-553b-4b5d-8760-b4b3d2547159=UNKNOWN_TOPIC_OR_PARTITION, hono.event.17c4c8ae-5150-4e23-99ae-4faeee9e0489=UNKNOWN_TOPIC_OR_PARTITION, hono.event.502e0622-bba2-49d3-b5ec-8a4fcca14241=UNKNOWN_TOPIC_OR_PARTITION, hono.event.6b283196-bca9-43a4-a799-ac3fe1c789b3=UNKNOWN_TOPIC_OR_PARTITION, hono.event.0dfaf89d-33e6-4138-b3c6-cd4e6474bf58=UNKNOWN_TOPIC_OR_PARTITION, hono.event.7046706f-127b-461e-b35a-09cfd091ff22=UNKNOWN_TOPIC_OR_PARTITION, hono.event.0ee960c1-b9cf-4dd8-95c6-01d13ac07b39=UNKNOWN_TOPIC_OR_PARTITION, hono.event.bbcd2940-6804-4b6e-bff0-13181104c116=UNKNOWN_TOPIC_OR_PARTITION, hono.event.87b6d19c-89c7-46ca-ae69-8b0ce470c58e=UNKNOWN_TOPIC_OR_PARTITION, hono.event.44966256-0c97-47e8-830d-92f76997b3a0=UNKNOWN_TOPIC_OR_PARTITION, hono.event.a98a395f-2202-4f1d-b84a-749e7e268e95=UNKNOWN_TOPIC_OR_PARTITION, hono.event.48cf99f4-df5b-435a-b252-4e05728b2898=UNKNOWN_TOPIC_OR_PARTITION, hono.event.e5e8049a-6a4e-4455-9df1-c685ae3c6427=UNKNOWN_TOPIC_OR_PARTITION, hono.event.136eaf85-4a55-4c6a-94b7-de6d6ec95da3=UNKNOWN_TOPIC_OR_PARTITION, hono.event.917664e5-edac-4f03-be88-98182d198ddc=UNKNOWN_TOPIC_OR_PARTITION, hono.event.b439f69f-eb8e-47dd-a010-ad5faba0fa0e=UNKNOWN_TOPIC_OR_PARTITION, hono.event.fb59448a-6238-4dd8-a517-c9b4f35fbaa0=UNKNOWN_TOPIC_OR_PARTITION, hono.event.5067b2cf-3ddd-4907-bb39-38cc8868ce9e=UNKNOWN_TOPIC_OR_PARTITION}
```

This might be a red herring though... I have not confirmed that a similar error appears every time the tests error out.

StFS commented:
correction. The error mentioned above does seem to appear every now and then in the infinispan based command router, but to be clear, the test runs do not error out. So the error found in the command router logs does not seem to be related to the test errors.

For reference, here is a similar error log line from the infinispan command router that happened while tests were running successfully:

```
2025-01-18 03:24:34,549 WARN  [org.apa.kaf.cli.NetworkClient] (kafka-producer-network-thread | event-2dbb1cf3-4f7a-46a3-97bf-53a37b95fd9e) [Producer clientId=event-2dbb1cf3-4f7a-46a3-97bf-53a37b95fd9e] Error while fetching metadata with correlation id 400 : {hono.event.33f6673d-fd5f-4f40-a163-79e475be9639=UNKNOWN_TOPIC_OR_PARTITION, hono.event.1c34ca87-2f55-4fe5-9529-b894d828f8ee=UNKNOWN_TOPIC_OR_PARTITION, hono.event.b9bc2de8-35b8-42d3-aa36-fa830184dd00=UNKNOWN_TOPIC_OR_PARTITION, hono.event.1b830943-a8d6-42f5-b5db-6f63df5915f8=UNKNOWN_TOPIC_OR_PARTITION, hono.event.a7b059d5-149f-4edf-964d-7bc27f00dae2=UNKNOWN_TOPIC_OR_PARTITION, hono.event.dbd5e165-5150-4f1c-aac5-cbed25fc8b79=UNKNOWN_TOPIC_OR_PARTITION, hono.event.8066ccfc-f915-4e4c-a5db-2bdc5327ffdf=UNKNOWN_TOPIC_OR_PARTITION, hono.event.6e620301-fbe7-4815-895d-1bfb8a497d93=UNKNOWN_TOPIC_OR_PARTITION, hono.event.02cf5756-1168-4fe6-940e-ca7b2b249ce1=UNKNOWN_TOPIC_OR_PARTITION, hono.event.59ca1b4f-4610-4f68-a278-e25e300d03ea=UNKNOWN_TOPIC_OR_PARTITION, hono.event.b98fdbb0-dad1-481f-a7f6-03cad74dba34=UNKNOWN_TOPIC_OR_PARTITION, hono.event.c16d2fc0-9159-46db-9eb6-011889e787c8=UNKNOWN_TOPIC_OR_PARTITION, hono.event.8b139e23-7f0f-4636-a6c7-22cf545c7db3=UNKNOWN_TOPIC_OR_PARTITION, hono.event.bd2b0553-4c59-4914-a750-07de11076ee2=UNKNOWN_TOPIC_OR_PARTITION, hono.event.8f5a34bc-2c72-4871-919a-958cc41467f0=UNKNOWN_TOPIC_OR_PARTITION, hono.event.7ab04826-f118-41f6-b3e2-bb2ed6d0ac33=UNKNOWN_TOPIC_OR_PARTITION, hono.event.a8c1efde-44b4-446a-8e4f-ff1293f2afec=UNKNOWN_TOPIC_OR_PARTITION, hono.event.6f0d1596-beba-4edc-97e3-5ef73dfb3b63=UNKNOWN_TOPIC_OR_PARTITION, hono.event.53211385-c399-4599-be0d-be0efe995373=UNKNOWN_TOPIC_OR_PARTITION, hono.event.5744d151-0ab5-4ee5-8f86-47a1f32dedbb=UNKNOWN_TOPIC_OR_PARTITION, hono.event.ba649b07-8f1d-4818-a966-8e094d9ec191=UNKNOWN_TOPIC_OR_PARTITION, hono.event.741bb2ca-6a74-4f30-a7bb-d4953f16302f=UNKNOWN_TOPIC_OR_PARTITION, hono.event.442b0067-e19d-4e9a-ba22-f0e21a908060=UNKNOWN_TOPIC_OR_PARTITION, hono.event.3eb43744-cd05-4f05-8200-b71d3ee48e5f=UNKNOWN_TOPIC_OR_PARTITION, hono.event.8e1c8c92-c797-429e-940b-58070441b3a7=UNKNOWN_TOPIC_OR_PARTITION, hono.event.a68682c1-f6de-47cb-abaa-22017c1ce13d=UNKNOWN_TOPIC_OR_PARTITION, hono.event.591e9c7a-6eec-45e1-9a31-1e06b330f66c=UNKNOWN_TOPIC_OR_PARTITION, hono.event.71398b9f-1407-4e9d-b5b2-b3a75cd4eb80=UNKNOWN_TOPIC_OR_PARTITION, hono.event.0f49b5e6-6196-4a34-a1ed-6e4555c3b447=UNKNOWN_TOPIC_OR_PARTITION, hono.event.6166a5c3-4d98-41da-9ed2-d1f5a595951c=UNKNOWN_TOPIC_OR_PARTITION, hono.event.7caae0e8-0b13-46cc-926c-98f955b8d7de=UNKNOWN_TOPIC_OR_PARTITION, hono.event.d86c9c63-0822-42ea-8f7d-a99d6e18644d=UNKNOWN_TOPIC_OR_PARTITION, hono.event.ff52b378-e18e-40bd-86ee-f657b45a7ed2=UNKNOWN_TOPIC_OR_PARTITION, hono.event.093ef358-c89a-4391-98bd-65a6d9209e52=UNKNOWN_TOPIC_OR_PARTITION, hono.event.f51bbed1-6d96-4084-a065-66aec2cdbc9c=UNKNOWN_TOPIC_OR_PARTITION, hono.event.6d7008b0-3811-48d8-b396-789973bb22db=UNKNOWN_TOPIC_OR_PARTITION, hono.event.2044a15c-0ed7-4afa-928b-7cfde63fd485=UNKNOWN_TOPIC_OR_PARTITION, hono.event.e78eba0d-fa11-47aa-b63e-c9fcaec9537e=UNKNOWN_TOPIC_OR_PARTITION, hono.event.321c29aa-044f-496b-ba69-24c45a7bd827=UNKNOWN_TOPIC_OR_PARTITION, hono.event.268d340f-98f2-4bb7-8289-644a0a180ce6=UNKNOWN_TOPIC_OR_PARTITION, hono.event.8b928a44-ec9c-4555-b56f-6b327cc719ef=UNKNOWN_TOPIC_OR_PARTITION, hono.event.44d72121-d8e5-4864-ae1f-9e802c988aa9=UNKNOWN_TOPIC_OR_PARTITION, hono.event.cb65cbae-d686-435a-9bf5-23b139672714=UNKNOWN_TOPIC_OR_PARTITION, hono.event.ce591207-6c3a-45ae-8410-64661625cda1=UNKNOWN_TOPIC_OR_PARTITION, hono.event.311f670d-18c4-42f9-9938-6ae6b32ddfab=UNKNOWN_TOPIC_OR_PARTITION, hono.event.9361b53d-a8e9-42d9-8e97-889ef56e3bd8=UNKNOWN_TOPIC_OR_PARTITION, hono.event.2e8374ed-8692-4567-aaab-5db4ac5173d5=UNKNOWN_TOPIC_OR_PARTITION, hono.event.942f0a7e-61c1-4632-9a34-f3e531e3485c=UNKNOWN_TOPIC_OR_PARTITION, hono.event.dd9e682e-8ee6-475f-bdd4-ab68704763fa=UNKNOWN_TOPIC_OR_PARTITION, hono.event.0ad4f733-47f1-41b0-997f-f7f21fe9e511=UNKNOWN_TOPIC_OR_PARTITION, hono.event.3e36a0f9-367f-42d1-81c7-8d31d8adecbf=UNKNOWN_TOPIC_OR_PARTITION, hono.event.4e0bdd48-ff19-4fe1-8487-4aa51b102ac9=UNKNOWN_TOPIC_OR_PARTITION}
```

StFS commented:
Sometimes the tests run successfully and sometimes they fail. There seems to be quite a variance in when they fail... sometimes they will fail on the first run but sometimes there will be 5-6 successful runs and then a run will fail. If a test run is repeated immediately after a failed test run (without rebooting the containers) the tests may run successfully a number of times again without failing, indicating that the command router service is not left in a permanently corrupted state after a failed test run.

Attached is a log of a failed test run.
[hono.test.error.log](https://github.com/user-attachments/files/18462762/hono.test.error.log)

StFS commented:
It seems that the failure happens eventually even if you only run that single test (although I've seen test from the Mqtt test fail as well. But this is sufficient to get the failure to appear eventually:

```
# Run once to start all the containers
mvn clean verify -Prun-tests,redis_cache -Dit.test=CommandAndControlMqttIT#testSendOneWayCommandSucceeds -Ddocker.keepRunning

# Run the tests repeatedly until a run fails
while mvn verify -Prun-tests,redis_cache,useRunningContainers -Dit.test=CommandAndControlMqttIT#testSendOneWayCommandSucceeds ; do :; done
```

This will eventually result in a failed test run and often (but not always) subsequent test runs will then fail consistently. Sometimes however, subsequent test runs pass successfully.

StFS committed:
Cleaning up code in redis config interceptor class

StFS committed:
Cleaning up code in redis config interceptor class

StFS committed:
Trying out bogus config value for redis server

StFS committed:
Trying out bogus config value for redis server

StFS committed:
Reverting bogus redis hosts value

StFS committed:
Reverting bogus redis hosts value

sophokles73 commented on services/command-router-redis/src/main/java/org/eclipse/hono/commandrouter/redis/config/RedisConfigInterceptor.java:
```diff
@@ -0,0 +1,19 @@
+/**
+ * TODO.
+ */
+public class RedisConfigInterceptor extends RelocateConfigSourceInterceptor {
```
FMPOV we should simply use the default Quarkus mechanism for configuring the connection to the Redis cache ...

StFS replied:
Do you mean using the quarkus redis configuration keys instead of re-mapping them under the hono namespace?

If that's acceptable I'd very much like that.

sophokles73 commented on services/command-router-redis/src/main/resources/META-INF/native-image/org.eclipse.hono/hono-service-command-router/native-image.properties:
```diff
@@ -0,0 +1,12 @@
+#
+# SPDX-License-Identifier: EPL-2.0
+
+Args = -H:+RunReachabilityHandlersConcurrently
```
Is this strictly necessary?

StFS replied:
I copied this from some other componnt (probably the infinispan command router) so I have not actually used this at all. I haven't run the native containers at all and honestly I don't really have any idea wht these reachability handler do.

sophokles73 commented on tests/pom.xml:
```diff
@@ -2227,7 +2276,7 @@
                   <removeVolumes>true</removeVolumes>
                   <removeNamePattern>**/hono-*-test:*</removeNamePattern>
-                  <stopNamePattern>hono-*-test-*,all-in-one-*,cp-*,postgres-*</stopNamePattern>
+                  <stopNamePattern>hono-*-test-*,all-in-one-*,cp-*,postgres-*,redis-*</stopNamePattern>
```
shouldn't the Redis container be covered by the `hono-*-test-*` pattern?

StFS replied:
well, I was copying the pre-existing postgres pattern there which started a postgresql container that does not use the hono-*-test-* naming pattern.

But if you think naming the redis container (that is, the Redis server itself, not the redis command router component being tested) something like hono-redis-test-1 then I can definitely make that happen.

sophokles73 reviewed (commented):
This looks pretty solid to me :-)
At first glance, I cannot see, why the tests should fail. I haven't run the tests yet but will first take another look at the integration test implementation. We sometimes have odd bugs regarding async behavior in them ...

sophokles73 commented:
@StFS I believe I have found the culprit. When the integration test sends the one-way commands at a high rate (as it does), the Command Router sometimes fails to query the Redis cache for the _command handling adapter instances_, which it needs to do for each command to be forwarded to a protocol adapter instance. The Jaeger trace includes the following stack trace:

```sh
io.vertx.core.http.ConnectionPoolTooBusyException: Connection pool reached max wait queue size of 40
	at io.vertx.core.net.impl.pool.SimpleConnectionPool$Acquire$6.run(SimpleConnectionPool.java:630)
	at io.vertx.core.net.impl.pool.Task.runNextTasks(Task.java:43)
	at io.vertx.core.net.impl.pool.CombinerExecutor.submit(CombinerExecutor.java:91)
	at io.vertx.core.net.impl.pool.SimpleConnectionPool.execute(SimpleConnectionPool.java:244)
	at io.vertx.core.net.impl.pool.SimpleConnectionPool.acquire(SimpleConnectionPool.java:639)
	at io.vertx.core.net.impl.pool.SimpleConnectionPool.acquire(SimpleConnectionPool.java:643)
	at io.vertx.redis.client.impl.RedisConnectionManager$RedisEndpoint.requestConnection(RedisConnectionManager.java:430)
	at io.vertx.core.net.impl.pool.Endpoint.getConnection(Endpoint.java:41)
	at io.vertx.core.net.impl.pool.ConnectionManager.getConnection(ConnectionManager.java:51)
	at io.vertx.core.net.impl.pool.ConnectionManager.getConnection(ConnectionManager.java:40)
	at io.vertx.redis.client.impl.RedisConnectionManager.getConnection(RedisConnectionManager.java:394)
	at io.vertx.redis.client.impl.RedisClient.connect(RedisClient.java:36)
	at io.vertx.redis.client.impl.BaseRedisClient.send(BaseRedisClient.java:42)
	at io.quarkus.redis.runtime.client.ObservableRedis.send(ObservableRedis.java:83)
	at io.vertx.redis.client.impl.RedisAPIImpl.send(RedisAPIImpl.java:56)
	at io.vertx.redis.client.RedisAPI.mget(RedisAPI.java:3791)
	at io.vertx.redis.client.RedisAPI_DVNloIyh16wAC4m7ssePIaeqbbc_Synthetic_ClientProxy.mget(Unknown Source)
	at org.eclipse.hono.deviceconnection.redis.client.RedisCache.getAll(RedisCache.java:183)
	at org.eclipse.hono.deviceconnection.common.CacheBasedDeviceConnectionInfo.getInstancesQueryingAllGatewaysFirst(CacheBasedDeviceConnectionInfo.java:320)
	at org.eclipse.hono.deviceconnection.common.CacheBasedDeviceConnectionInfo.getCommandHandlingAdapterInstances(CacheBasedDeviceConnectionInfo.java:296)
	at org.eclipse.hono.commandrouter.impl.CommandTargetMapperImpl.lambda$getTargetGatewayAndAdapterInstance$2(CommandTargetMapperImpl.java:96)
	at io.vertx.core.impl.future.Composition.onSuccess(Composition.java:38)
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:66)
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
	at io.vertx.core.impl.future.Composition$1.onSuccess(Composition.java:62)
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:66)
	at io.vertx.core.impl.future.SucceededFuture.addListener(SucceededFuture.java:88)
	at io.vertx.core.impl.future.Composition.onSuccess(Composition.java:43)
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:66)
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
	at io.vertx.core.impl.future.Mapping.onSuccess(Mapping.java:40)
	at io.vertx.core.impl.future.FutureImpl$ListenerArray.onSuccess(FutureImpl.java:310)
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:66)
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
	at io.vertx.core.impl.future.Mapping.onSuccess(Mapping.java:40)
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:66)
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
	at io.vertx.core.impl.future.Composition$1.onSuccess(Composition.java:62)
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:66)
	at io.vertx.core.impl.future.SucceededFuture.addListener(SucceededFuture.java:88)
	at io.vertx.core.impl.future.Composition.onSuccess(Composition.java:43)
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:66)
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
	at io.vertx.core.impl.future.Composition$1.onSuccess(Composition.java:62)
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:66)
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
	at io.vertx.core.impl.future.Mapping.onSuccess(Mapping.java:40)
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:66)
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
	at io.vertx.core.impl.future.Composition$1.onSuccess(Composition.java:62)
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:66)
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
	at io.vertx.core.impl.future.PromiseImpl.onSuccess(PromiseImpl.java:49)
	at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:41)
	at org.eclipse.hono.client.amqp.RequestResponseClient.handleResponse(RequestResponseClient.java:368)
	at org.eclipse.hono.client.amqp.connection.impl.HonoConnectionImpl.lambda$createReceiver$13(HonoConnectionImpl.java:688)
	at io.vertx.proton.impl.ProtonReceiverImpl.onDelivery(ProtonReceiverImpl.java:236)
	at io.vertx.proton.impl.ProtonTransport.handleSocketBuffer(ProtonTransport.java:168)
	at io.vertx.core.net.impl.NetSocketImpl.lambda$new$1(NetSocketImpl.java:104)
	at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer.java:255)
	at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:134)
	at io.vertx.core.net.impl.NetSocketImpl$DataMessageHandler.handle(NetSocketImpl.java:412)
	at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:328)
	at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:321)
	at io.vertx.core.net.impl.NetSocketImpl.handleMessage(NetSocketImpl.java:388)
	at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:159)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:801)
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:509)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:407)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Unknown Source)
```

I have increased the connection pool:
```yaml
  commandRouter:
    amqp:
      insecurePortEnabled: true
      insecurePortBindAddress: "0.0.0.0"
    cache:
      redis:
        hosts: "redis://redis:6379"
        max-pool-size: 8
        max-pool-waiting: 40
```

but this did not (yet) fix the issue. I can also hardly believe that just running a few queries already brings Redis to its limits. Maybe you want to investigate a little further ...

StFS committed:
Merge remote-tracking branch 'origin/master' into create_redis_command_router_module

StFS committed:
Merge remote-tracking branch 'origin/master' into create_redis_command_router_module

StFS committed:
Fix Redis transactions failing in connection-less mode

Use dedicated connections for Redis operations that require
WATCH/MULTI/EXEC transactions, since the pooled RedisAPI does
not support stateful transactional commands. Also remove the
api.close() call from stop() as the RedisAPI lifecycle is
managed by the Quarkus CDI container.

StFS committed:
Fix Redis transactions failing in connection-less mode

Use dedicated connections for Redis operations that require
WATCH/MULTI/EXEC transactions, since the pooled RedisAPI does
not support stateful transactional commands. Also remove the
api.close() call from stop() as the RedisAPI lifecycle is
managed by the Quarkus CDI container.

StFS committed:
Fix property name for kafka image

sophokles73 committed:
[#3722] Bump Quarkus to 3.27.2

sophokles73 committed:
[#3726] Remove obsolete Nightly Jenkins job definition

There currently is no need to publish SNAPSHOTS of the Java artifacts
to a Maven repository on a nightly basis, as the Hono project does not
have any consumers depending on such snapshots. Therefore, the
nightly Jenkins job definition is obsolete and can be removed.

sophokles73 committed:
[#3729] Put dev docs into distinct language folder

The dev version of the documentation has not been built correctly
anymore since the update to Hugo 0.123.8.
This seems to have been caused by the fact that the "dev" version
content has not been put into its own "content/dev" subfolder.

sophokles73 committed:
[#3726] Migrate Sonar scan job to GitHub Actions

Replaced existing Jenkins pipeline for Sonar scans with a GitHub
Actions workflow.

sophokles73 committed:
[#3726] Fix job name in Sonar scan workflow

The name of the job in the Sonar scan workflow was not correct, which
made it difficult to identify the job in the GitHub Actions UI.
This commit updates the job name for better clarity.

Also updated JVM version to 21 and changed Maven command to run in
batch mode and adapted the checkout step to fetch the full history in
order to allow Sonar to process blame information.

sophokles73 committed:
Fix homepage links to use eclipse.dev domain

sophokles73 committed:
Remove unmaintained/obsolete web site build script for Windows

With the advent of WSL there does not seem to be much need/use for the
Windows build script (site/build-site.bat) anymore.
Removed it to reduce maintenance burden.

Also fixed *nix build script to handle paths with spaces and updated
the Maven exec plugin version in the POM to a more recent one.

sophokles73 committed:
[#3726] Remove obsolete Website Jenkins Pipeline

The Jenkins pipeline for the Hono website is no longer needed, as the
website is now built and deployed using GitHub Actions.

With the advent of WSL there does not seem to be much need/use for the
Windows build script anymore.
Removed it to reduce maintenance burden.

sophokles73 committed:
Update Hono website URL in all POM, source and markdown files

Removed duplicate/unnecessary URLs from all POM files.

sophokles73 committed:
Include website build in CI workflow

Make sure to also build the website in the CI workflow, so that we can
catch any issues with the website build early on.

sophokles73 committed:
[#3726] Migrate Jenkins release pipelines to GitHub Actions

Replaced the existing Jenkins based jobs with GitHub Actions workflows
to automate releases to Maven Central.

Also replaced the Sonatype release plugin with the Maven Njord plugin to
publish artifacts to Sonatype Central. The Njord plugin allows deploying
artifacts first to a local store in the filesystem before then
publishing them to whatever repository is being used.

sophokles73 committed:
Do not build website when doing Sonar scan to avoid hugo related errors

The Sonar scan had failed due a missing hugo binary. Building the
site does not provide any insights in this context so we simply disable
the website build when running the Sonar scan.

sophokles73 committed:
Prepare Hono 2.7.0 release

Updated ActiveMQ Artemis Broker image to latest version to fix a
startup issue.

sophokles73 committed:
Install Hugo before release build

sophokles73 committed:
Check out submodules recursively for release and publish workflow

sophokles73 committed:
Use standard docker.io registry for our Docker images

Also login to Docker Hub in the release workflow before building and
pushing our container images, so that we can push multi-platform images
using Buildx.

sophokles73 committed:
Use builder created by setup-buildx action for multiarch docker build

eclipse-hono-bot committed:
Release 2.7.0

eclipse-hono-bot committed:
Bump version to 2.8.0-SNAPSHOT

sophokles73 committed:
Update docs for 2.7.x branch

Also updated Downloads page to refer to GitHub releases instead of
EF download server.

sophokles73 committed:
Add JVM based command line client to release artifacts on GitHub

StFS committed:
Merge remote-tracking branch 'upstream' into create_redis_command_router_module

StFS committed:
Update version to 2.8.0-SNAPSHOT

StFS committed:
Update version to 2.8.0-SNAPSHOT

StFS committed:
Fix port issue on Kafka between tests runs when using docker.keepRunning and useRunningContainers

StFS committed:
Fix port issue on Kafka between tests runs when using docker.keepRunning and useRunningContainers

StFS committed:
Got redis pool metrics working

StFS committed:
Got redis pool metrics working

StFS commented:
Finally got redis metrics working properly and was able to monitor them (1 second interval) while running the integration tests from start and until the tests failed.
<img width="3486" height="6288" alt="grafana-redis-metrics-commandrouter" src="https://github.com/user-attachments/assets/276a4e3a-1ad9-4017-9bb0-2601061467ea" />

StFS commented on client-device-connection-redis/src/main/java/org/eclipse/hono/deviceconnection/redis/client/RedisCache.java:
```diff
@@ -0,0 +1,195 @@
+                        }
+                        futures.add(api.set(params));
+                    });
+                    return Future.all(Collections.unmodifiableList(futures));
```
I'm refactoring this code a bit so I'll ask for a re-review once that's done

StFS commented on client-device-connection-redis/src/main/java/org/eclipse/hono/deviceconnection/redis/client/RedisCache.java:
```diff
@@ -0,0 +1,195 @@
+                })
+                .compose(ignored -> api.exec())
+                // null reply means transaction aborted
+                .map(Objects::nonNull)
```
I'll look into that.

StFS commented on client-device-connection-redis/src/main/java/org/eclipse/hono/deviceconnection/redis/client/RedisCache.java:
```diff
@@ -0,0 +1,195 @@
+        Objects.requireNonNull(key);
+
+        return api.get(key)
+                .compose(value -> Future.succeededFuture(String.valueOf(value)));
```
Will consider this during refactoring.

StFS commented on client-device-connection-redis/src/main/java/org/eclipse/hono/deviceconnection/redis/client/RedisCache.java:
```diff
@@ -0,0 +1,195 @@
+                        // key does not exist
+                        return Future.succeededFuture(false);
+                    }
+                    if (String.valueOf(response).equals(value)) {
```
will take this into account

StFS committed:
Raise Redis connection pool wait queue to absorb command bursts

The vert.x Redis client acquires a pooled connection per command, so a
burst of commands to route (e.g. one Kafka poll batch, each command
needing a cache lookup) easily exceeds the pool's capacity. Acquisitions
beyond max-pool-waiting (Quarkus default 24) are rejected with
ConnectionPoolTooBusyException, silently failing command delivery and
causing the sporadic CommandAndControlAmqpIT/CommandAndControlMqttIT
failures (confirmed via redis_pool_rejected_total and Jaeger traces).

Ship max-pool-waiting=512 as a module default: a deep wait queue smooths
the spike at a few milliseconds of queueing latency per lookup instead
of dropping commands. Validated with 12 consecutive integration test
loop runs (~16k Redis commands, zero pool rejections).

StFS committed:
Fix checkstyle violations in RedisCache from stray static import

StFS committed:
Add Redis cache contract tests running against a containerized Redis

StFS committed:
Let RedisCache.get succeed with null for missing keys as required by the Cache contract

StFS committed:
Replace WATCH/MULTI based compare-and-delete with an atomic Lua script to avoid poisoning pooled connections

StFS committed:
Replace MULTI based putAll-with-lifespan with an atomic Lua script and guard empty maps

StFS committed:
Simplify RedisCache to depend on RedisAPI only and document the class

StFS committed:
Fail fast at startup if the Redis server does not permit Lua scripting

StFS committed:
Ensure the restricted ACL test user is always removed even if the scripting probe unexpectedly succeeds

StFS committed:
Document the Redis Lua scripting requirement of the device connection client

StFS committed:
Log Redis cache startup failures and make the no-expiry guarantee observable in tests

StFS committed:
Document the Redis command router and flag it as experimental

Bring the Redis variant of the command router up to the same
documentation standard as the Infinispan variant:

- Replace the TODO placeholders in RedisConfigInterceptor with a proper
  license header and Javadoc explaining the quarkus.redis to
  hono.commandRouter.cache.redis property relocation.
- Correct the copy-pasted class Javadoc in RedisDeviceConnectionInfoProducer
  which incorrectly described an in-memory/embedded cache option.
- Add a Redis cache section to the command router admin guide, covering
  the connection properties, the Lua/EVAL scripting requirement and an
  experimental-feature warning.
- Add an application-dev.properties for quarkus:dev parity with the
  Infinispan module and restore the Maven placeholders in the banner so
  the running service prints its name and version.
- Add an experimental notice to the Redis device connection client README.

StFS committed:
Rename getAll loop variable from i to value for clarity

StFS committed:
Remove unused native-image.properties from command-router-redis

StFS committed:
Make Redis pool sizing explicit in command router IT config

StFS committed:
Restore kafka.port.properties write-project-properties execution

Reverts an unintended change on this branch that removed the
properties-maven-plugin execution present on master, keeping the PR diff
limited to the Redis command router work.

StFS committed:
Revert unrelated Jaeger memory and Kafka port binding changes

Restores the master values that were changed on this branch but are
unrelated to the Redis command router work, keeping the PR diff focused.
